File: greedy/jump_game/jump_game1.go
package jumpgame

// 用例通过数 169/170， bad case [4,2,0,0,1,1,4,4,4,0,4,0]
func canJumpWrongWay(nums []int) bool {
	n := len(nums) - 1
	for i := 0; i <= n; {
		steps := nums[i]
		// [i, steps+i]这个区间内的最大数
		maxSteps := 0
		// 这里需要保存下i的值
		tmp := i
		for j := tmp + 1; j <= steps+tmp; j++ {
			// 这里要>=，即使是等于，也尽量往后走
			if nums[j] >= maxSteps {
				maxSteps = nums[j]
				i = j
			}
		}
		if maxSteps == 0 {
			return false
		}
		if maxSteps+i >= n {
			return true
		}
	}
	return false
}

// 贪心的策略其实是一步步的走，不断统计最大能够覆盖的范围
func canJump(nums []int) bool {
	if len(nums) <= 1 {
		return true
	}
	var cover int
	for i := 0; i <= cover; i++ {
		if cover < nums[i]+i {
			cover = nums[i] + i
		}
		if cover >= len(nums)-1 {
			return true
		}
	}
	return false
}

File: skill/jump_game/JumpGame1.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-11-17 16:18:48
 * @LastEditTime: 2023-02-18 15:37:36
 */
package skill.jump_game;

/**
 * leetcode 55 middle 跳跃游戏
 * 
 * 给定一个非负整数数组 nums ，你最初位于数组的 第一个下标。
 * 数组中的每个元素代表你在该位置可以跳跃的最大长度。
 * 判断你是否能够到达最后一个下标。
 */
public class JumpGame1 {
    // 错误的写法，bad case [3,2,1,0,4]
    @Deprecated
    public boolean canJump(int[] nums) {
        int n = nums.length;
        boolean[] canReach = new boolean[n];
        
        for (int i = 0; i < nums.length; i++) {
            if (i + nums[i] >= n) {
                return true;
            }
            for (int j = i; j < i + nums[i]; j++) {
                canReach[j] = true;
            }
        }

        return canReach[n - 1];
    }

    public boolean canJumpRightWay(int[] nums) {
        int n = nums.length;
        int maxReach = 0;

        for (int i = 0; i < nums.length; i++) {
            maxReach = Math.max(maxReach, i + nums[i]);
            if (maxReach >= n - 1) {
                return true;
            }
        }

        return false;
    }
}
